### PR TITLE
upgrade to Scala 3

### PR DESCRIPTION
**Pull request by mpollmeier**
The biggest visible change is the repl, which is a complete reimplementation. 
Ammonite is no longer in our dependency tree, and I factored out the core repl
functionality into a separate build: https://github.com/mpollmeier/scala-repl-pp/

The repl options and usage are identical though, apart from an additional `--dependency` 
option which I'll document on joern-website once this is merged. 

The main downstream builds are ready to get upgraded as well. 

Note: the code is ready to get reviewed, however we won't merge anything until we're 
happy with the feedback from the manual user testing.

### CODE REVIEW

File: joern-cli/src/main/resources/scripts/c/malloc-overflow.sc [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
-    }
-    .l
+    }.code
+
+  val expected = Set(
+    "malloc(sizeof(int) * 42)",
+    "malloc(sizeof(int) * 3)",
+    "malloc(sizeof(int) + 55)"
+  )
+  assertContains("calls", calls, expected)
```
Comment (ml86) [inline joern-cli/src/main/resources/scripts/c/malloc-overflow.sc:20]: Did this turn into some kind of test? Before we did not assert anything. Did the use of this script change. Some question applies to the other `.sc` files.

File: benchmarks/build.sbt [hunk 2]
```diff
@@ -16,12 +14,8 @@ Compile / doc / sources ~= (_ filter (_ => false))
...
 // we want to consume this from a java8 build
 compile / javacOptions ++= Seq("--release", "8")
-scalacOptions ++= Seq() ++ (
-  CrossVersion.partialVersion(scalaVersion.value) match {
-    case Some((3, _)) => Seq("-Xtarget:8")
-    case _            => Seq("-target:jvm-1.8")
-  }
-)
+
+scalacOptions += "-Xtarget:8"
```
Comment (ml86) [inline benchmarks/build.sbt:18]: Cant we set this for all subprojects in the main `build.sbt`?

File: joern-cli/src/main/resources/scripts/c/malloc-overflow.sc (hunk 1 above)
Comment (ml86) [reply joern-cli/src/main/resources/scripts/c/malloc-overflow.sc:20]: I just noticed that those expectation came from `RunTestScripts.scala`. Why did you change the setup? Having the tests in the actual implementation seems strange.

File: macros/src/main/scala-3/io/joern/macros/QueryMacros.scala [hunk 3]
```diff
@@ -10,13 +10,14 @@ import scala.quoted.{Expr, Quotes}
...
+    travExpr: Expr[Cpg => Traversal[_ <: StoredNode]]
+  )(using quotes: Quotes): Expr[TraversalWithStrRep] = {
     import quotes.reflect._
-    val pos = travExpr.asTerm.pos
+    val pos  = travExpr.asTerm.pos
     val code = Position(pos.sourceFile, pos.start, pos.end).sourceCode.getOrElse("N/A")
-    '{TraversalWithStrRep(${travExpr}, ${Expr(code)})}
+    '{ TraversalWithStrRep(${ travExpr }, ${ Expr(code) }) }
   }
 }
```
Comment (ml86) [inline macros/src/main/scala-3/io/joern/macros/QueryMacros.scala:23]: There is a scala 2 version of the `QueryMacros` implementation which i would have expected to be deleted. The file location can than also be adjusted (no more `scala-3` in path).

File: benchmarks/build.sbt (hunk 2 above)
Comment (mpollmeier) [reply benchmarks/build.sbt:18]: good idea, done

File: macros/src/main/scala-3/io/joern/macros/QueryMacros.scala (hunk 3 above)
Comment (mpollmeier) [reply macros/src/main/scala-3/io/joern/macros/QueryMacros.scala:23]: good point, done

**Pull request closed without merging**